ext/vulnsrc/ubuntu/ubuntu.go

```diff
@@ -13,369 +13,406 @@
+		return resp, err
 	}
+	defer r.Body.Close()
 
-	// Set the updaterFlag to equal the commit processed.
-	resp.FlagName = updaterFlag
-	resp.FlagValue = commit
+	var ovalFiles []string
+	var generationTimes []int64
 
-	// Short-circuit if there have been no updates.
-	if commit == dbCommit {
-		log.WithField("package", "ubuntu").Debug("no update")
-		return
-	}
+	scanner := bufio.NewScanner(r.Body)
+	for scanner.Scan() {
+		line := scanner.Text()
+		r := ubuntuOvalFileRegexp.FindStringSubmatch(line)
+		if len(r) != 2 {
+			continue
+		}
 
-	// Get the list of vulnerabilities that we have to update.
-	var modifiedCVE map[string]struct{}
-	modifiedCVE, err = collectModifiedVulnerabilities(commit, dbCommit, u.repositoryLocalPath)
-	if err != nil {
-		return
+		ovalFile := ovalURI + ubuntuOvalFilePrefix + r[1] + ".cve.oval.xml.bz2"
+		log.WithFields(log.Fields{
+			"ovalFile": ovalFile,
+			"updater":  "Ubuntu Linux",
+		}).Debug("file to check")
+
+		// Do not fetch the entire file to get the value of the
+		// creation time. Rely on the "latest modified time"
+		// value of the file hosted on the remote server.
+		timestamp, err := getLatestModifiedTime(ovalFile)
+		if err != nil {
+			log.WithError(err).WithField("ovalFile", ovalFile).Warning("Ignoring OVAL file")
+		}
+
+		if timestamp > latestOval {
+			ovalFiles = append(ovalFiles, ovalFile)
+		}
 	}
 
-	// Get the list of vulnerabilities.
-	resp.Vulnerabilities, resp.Notes, err = collectVulnerabilitiesAndNotes(u.repositoryLocalPath, modifiedCVE)
-	if err != nil {
-		return
+	for _, oval := range ovalFiles {
+		log.WithFields(log.Fields{
+			"ovalFile": oval,
+			"updater":  "Ubuntu Linux",
+		}).Debug("downloading")
+		// Download the oval XML file.
+		r, err := http.Get(oval)
+		if err != nil {
+			log.WithError(err).Error("could not download Ubuntu update list")
+			return resp, commonerr.ErrCouldNotDownload
+		}
+		defer r.Body.Close()
+
+		// Parse the XML.
+		vs, generationTime, err := parseOval(bzip2.NewReader(r.Body))
+		if err != nil {
+			return resp, err
+		}
+		generationTimes = append(generationTimes, generationTime)
+
+		// Collect vulnerabilities.
+		resp.Vulnerabilities = append(resp.Vulnerabilities, vs...)
 	}
 
-	// The only notes we take are if we encountered unknown Ubuntu release.
-	// We don't want the commit to be considered as managed in that case.
-	if len(resp.Notes) != 0 {
-		resp.FlagValue = dbCommit
+	// Set the flag if we found anything.
+	if len(generationTimes) > 0 {
+		resp.FlagName = updaterFlag
+		resp.FlagValue = strconv.FormatInt(latest(generationTimes), 10)
+	} else {
+		log.WithField("package", "Ubuntu Linux").Debug("no update")
 	}
 
-	return
+	return resp, nil
 }
 
-func (u *updater) Clean() {
-	if u.repositoryLocalPath != "" {
-		os.RemoveAll(u.repositoryLocalPath)
+// Get the latest modification time of a remote file
+// expressed as unix time
+func getLatestModifiedTime(url string) (int64, error) {
+	resp, err := http.Head(url)
```

Comment:
@alexmurray  - believe you still want to defer close of the response body here. https://golang.org/pkg/net/http/ the docs groups the "head" method with the advice to close the body.

Reply:
This was copied from the same function in the suse vulnsrc - if this is the case that should also be fixed too.

Reply:
@alexmurray sounds like the right thing to do :+1:

Reply:
Would  you prefer a separate PR for this or should I bundle it with this one?

Reply:
Separate pr sounds good to me.